### PR TITLE
fix: per datapoint error handling

### PR DESCRIPTION
**Pull request by polomani**
Screenshot 2020-05-27 at 14 53 31

### CODE REVIEW

File: src/datasource.ts [hunk 1]
```diff
@@ -356,28 +356,37 @@ export function getRange(range: TimeRange): Tuple<number> {
...
+      (datapoints as [])
+        .map(({ error }) => error)
+        .filter(Boolean)
+        .forEach(error => {
+          calculationErrors.add(error);
+        });
+    });
+
+    calculationErrors.forEach(message => {
+      appEvents.emit(datapointsWarningEvent, { refId, warning: message });
```
Comment (mrtraser) [inline src/datasource.ts:389]: mb we can send only one event here but with the array of warnings?

File: src/datasource.ts [hunk 2]
```diff
@@ -356,28 +356,37 @@ export function getRange(range: TimeRange): Tuple<number> {
...
+    const hasMorePoints = items.some(({ datapoints }) => datapoints.length >= limit);
+    const warning =
+      'Datapoints limit was reached, so not all datapoints may be shown. Try increasing the granularity, or choose a smaller time range.';
+
+    if (hasMorePoints) {
+      appEvents.emit(datapointsWarningEvent, { refId, warning });
     }
-  );
+
+    const calculationErrors = new Set();
```
Comment (mrtraser) [inline src/datasource.ts:378]: can we split logic into separate functions (`checkLimits`, `checkCalculationErrors` ) ?

File: src/datasource.ts (hunk 1 above)
Comment (polomani) [reply src/datasource.ts]: fixed, but I've sent it as one string
should be ok as well

File: src/datasource.ts (hunk 2 above)
Comment (polomani) [reply src/datasource.ts]: fixed

**Pull request merged**